### PR TITLE
Opt into Ruby style for thoughtbot

### PR DESCRIPTION
**Pull request by salbertson**
This change is meant to enable cops that thoughtbot projects should follow and disable the rest.

### CODE REVIEW

File: config/style_guides/thoughtbot/ruby.yml [hunk 1]
```diff
@@ -1,260 +1,1063 @@
...
+  - "**/Gemfile"
+  - "**/Rakefile"
+  - "**/Capfile"
+  - "**/Guardfile"
+  - "**/Podfile"
+  - "**/Thorfile"
+  - "**/Vagrantfile"
+  - "**/Berksfile"
+  - "**/Cheffile"
+  - "**/Vagabondfile"
```
Comment (gylaz) [inline config/style_guides/thoughtbot/ruby.yml]: Is this what RuboCop enables by default? Seems kinda noisy.

**Comment by gylaz** [comment]
**L**e**G**i**T****M**

File: config/style_guides/thoughtbot/ruby.yml (hunk 1 above)
Comment (salbertson) [inline config/style_guides/thoughtbot/ruby.yml]: Yes, feel free to comment on the stuff that should be removed for thoughtbot projects.

**Comment by salbertson** [comment]
@gylaz huh?

File: config/style_guides/thoughtbot/ruby.yml (hunk 1 above)
Comment (gylaz) [inline config/style_guides/thoughtbot/ruby.yml]: I would say all of these except `Rakefile`, `Gemfile`, `**/*.gemspec`, and `**/*.rake`.

**Comment by gylaz** [comment]
LGTM

**Comment by salbertson** [comment]
@gylaz, I don't get it, all the Cops are disabled.

**Comment by gylaz** [comment]
@salbertson I'm confused about what you are confused about. LGTM, means "Looks good to me".

**Comment by jferris** [comment]
Is the idea with this pull request that we're starting with a file where everything is disabled, and we'll comment on the lines we think should be enabled based on our style guide?

**Comment by gylaz** [comment]
Do we have a reason to believe the current approach isn't working?

**Comment by jferris** [comment]
> Do we have a reason to believe the current approach isn't working?

The issue is that new rules are activated without input from the team, so occasionally Hound just shows up and starts commenting about odd things on pull requests.

Here's a good example: https://github.com/thoughtbot/upcase/pull/1228

> Assignment Branch Condition size for stub_trail is too high. [15.03/15]

We don't want Hound to do complexity checks; we have other tools for that.

> Rename `have_completed_text` to `have_completed_text?`.

We also consider it *good* practice to name matchers that way, so the second comment is sort of harmful.

We have a group buy-in process for our style guide. When Hound starts making up new rules, the process starts to break down.

**Comment by salbertson** [comment]
@jferris, yes, please :+1: any rule you would like enabled and I'll make the change.

File: config/style_guides/thoughtbot/ruby.yml [hunk 2]
```diff
@@ -1,260 +1,1063 @@
...
+  Description: Enforces consistent use of `Object#is_a?` or `Object#kind_of?`.
+  Enabled: false
+  EnforcedStyle: is_a?
+  SupportedStyles:
+  - is_a?
+  - kind_of?
+Style/CollectionMethods:
+  Description: Preferred collection methods.
+  StyleGuide: https://github.com/bbatsov/ruby-style-guide#map-find-select-reduce-size
   Enabled: false
```
Comment (gylaz) [inline config/style_guides/thoughtbot/ruby.yml]: Enable

File: config/style_guides/thoughtbot/ruby.yml [hunk 3]
```diff
@@ -1,260 +1,1063 @@
...
+  Description: Check indentation of private/protected visibility modifiers.
+  StyleGuide: https://github.com/bbatsov/ruby-style-guide#indent-public-private-protected
+  Enabled: false
+  EnforcedStyle: indent
+  SupportedStyles:
+  - outdent
+  - indent
+Style/AlignHash:
+  Description: Align the elements of a hash literal if they span more than one line.
+  Enabled: false
```
Comment (gylaz) [inline config/style_guides/thoughtbot/ruby.yml]: Enable

File: config/style_guides/thoughtbot/ruby.yml [hunk 4]
```diff
@@ -1,260 +1,1063 @@
...
+  EnforcedLastArgumentHashStyle: always_inspect
+  SupportedLastArgumentHashStyles:
+  - always_inspect
+  - always_ignore
+  - ignore_implicit
+  - ignore_explicit
+Style/AlignParameters:
+  Description: Align the parameters of a method call if they span more than one line.
+  StyleGuide: https://github.com/bbatsov/ruby-style-guide#no-double-indent
+  Enabled: false
```
Comment (gylaz) [inline config/style_guides/thoughtbot/ruby.yml]: Enable

File: config/style_guides/thoughtbot/ruby.yml [hunk 5]
```diff
@@ -1,260 +1,1063 @@
...
+  EnforcedStyle: with_first_parameter
+  SupportedStyles:
+  - with_first_parameter
+  - with_fixed_indentation
+Style/AndOr:
+  Description: Use &&/|| instead of and/or.
+  StyleGuide: https://github.com/bbatsov/ruby-style-guide#no-and-or-or
+  Enabled: false
```
Comment (gylaz) [inline config/style_guides/thoughtbot/ruby.yml]: I think we want this?

File: config/style_guides/thoughtbot/ruby.yml [hunk 6]
```diff
@@ -1,260 +1,1063 @@
...
+  EnforcedStyle: always
+  SupportedStyles:
+  - always
+  - conditionals
+Style/BarePercentLiterals:
+  Description: Checks if usage of %() or %Q() matches configuration.
+  StyleGuide: https://github.com/bbatsov/ruby-style-guide#percent-q-shorthand
+  Enabled: false
```
Comment (gylaz) [inline config/style_guides/thoughtbot/ruby.yml]: Enable

File: config/style_guides/thoughtbot/ruby.yml [hunk 7]
```diff
@@ -1,260 +1,1063 @@
...
+  EnforcedStyle: bare_percent
+  SupportedStyles:
+  - percent_q
+  - bare_percent
+Style/BracesAroundHashParameters:
+  Description: Enforce braces style around hash parameters.
+  Enabled: false
```
Comment (gylaz) [inline config/style_guides/thoughtbot/ruby.yml]: I think this is just idiomatic ruby.

File: config/style_guides/thoughtbot/ruby.yml [hunk 8]
```diff
@@ -1,260 +1,1063 @@
...
+  EnforcedStyle: no_braces
+  SupportedStyles:
+  - braces
+  - no_braces
+  - context_dependent
+Style/CaseIndentation:
+  Description: Indentation of when in a case/when/[else/]end.
+  StyleGuide: https://github.com/bbatsov/ruby-style-guide#indent-when-to-case
+  Enabled: false
```
Comment (gylaz) [inline config/style_guides/thoughtbot/ruby.yml]: Idiomatic ruby?

File: config/style_guides/thoughtbot/ruby.yml [hunk 9]
```diff
@@ -1,260 +1,1063 @@
...
+  IndentWhenRelativeTo: case
+  SupportedStyles:
+  - case
+  - end
+  IndentOneStep: false
+Style/ClassAndModuleChildren:
+  Description: Checks style of children classes and modules.
+  Enabled: false
```
Comment (gylaz) [inline config/style_guides/thoughtbot/ruby.yml]: Community best practice?

File: config/style_guides/thoughtbot/ruby.yml [hunk 10]
```diff
@@ -1,260 +1,1063 @@
...
+  EnforcedStyle: nested
+  SupportedStyles:
+  - nested
+  - compact
+Style/ClassCheck:
+  Description: Enforces consistent use of `Object#is_a?` or `Object#kind_of?`.
+  Enabled: false
```
Comment (gylaz) [inline config/style_guides/thoughtbot/ruby.yml]: We don't have a guide for this...

File: config/style_guides/thoughtbot/ruby.yml [hunk 11]
```diff
@@ -1,260 +1,1063 @@
...
+Style/CollectionMethods:
+  Description: Preferred collection methods.
+  StyleGuide: https://github.com/bbatsov/ruby-style-guide#map-find-select-reduce-size
   Enabled: false
-
-ActionFilter:
+  PreferredMethods:
+    collect: map
+    collect!: map!
+    find: detect
```
Comment (gylaz) [inline config/style_guides/thoughtbot/ruby.yml]: `detect: find`

File: config/style_guides/thoughtbot/ruby.yml [hunk 12]
```diff
@@ -1,260 +1,1063 @@
...
+    find_all: select
+    reduce: inject
```
Comment (gylaz) [inline config/style_guides/thoughtbot/ruby.yml]: `inject: reduce`

File: config/style_guides/thoughtbot/ruby.yml [hunk 13]
```diff
@@ -1,260 +1,1063 @@
...
+  Keywords:
+  - TODO
+  - FIXME
+  - OPTIMIZE
+  - HACK
+  - REVIEW
+Style/DotPosition:
+  Description: Checks the position of the dot in multi-line method calls.
+  StyleGuide: https://github.com/bbatsov/ruby-style-guide#consistent-multi-line-chains
   Enabled: false
```
Comment (gylaz) [inline config/style_guides/thoughtbot/ruby.yml]: Enable

File: config/style_guides/thoughtbot/ruby.yml [hunk 14]
```diff
@@ -1,260 +1,1063 @@
...
-
-Alias:
+  EnforcedStyle: trailing
+  SupportedStyles:
+  - leading
+  - trailing
+Style/EmptyLineBetweenDefs:
+  Description: Use empty lines between defs.
+  StyleGuide: https://github.com/bbatsov/ruby-style-guide#empty-lines-between-methods
+  Enabled: false
```
Comment (gylaz) [inline config/style_guides/thoughtbot/ruby.yml]: Enable

File: config/style_guides/thoughtbot/ruby.yml [hunk 15]
```diff
@@ -1,260 +1,1063 @@
...
+  AllowAdjacentOneLineDefs: false
+Style/EmptyLinesAroundBlockBody:
+  Description: Keeps track of empty lines around block bodies.
+  Enabled: false
```
Comment (gylaz) [inline config/style_guides/thoughtbot/ruby.yml]: Enable

File: config/style_guides/thoughtbot/ruby.yml [hunk 16]
```diff
@@ -1,260 +1,1063 @@
...
+  EnforcedStyle: no_empty_lines
```
Comment (gylaz) [inline config/style_guides/thoughtbot/ruby.yml]: `empty_lines`

File: config/style_guides/thoughtbot/ruby.yml [hunk 17]
```diff
@@ -1,260 +1,1063 @@
...
+  SupportedStyles:
+  - empty_lines
+  - no_empty_lines
+Style/EmptyLinesAroundClassBody:
+  Description: Keeps track of empty lines around class bodies.
+  Enabled: false
```
Comment (gylaz) [inline config/style_guides/thoughtbot/ruby.yml]: Enable? I'm pretty sure we don't recommend empty lines around block bodies, and thus Hound should comment to remove them.

File: config/style_guides/thoughtbot/ruby.yml [hunk 18]
```diff
@@ -1,260 +1,1063 @@
...
+  EnforcedStyle: no_empty_lines
+  SupportedStyles:
+  - empty_lines
+  - no_empty_lines
+Style/EmptyLinesAroundModuleBody:
+  Description: Keeps track of empty lines around module bodies.
+  Enabled: false
```
Comment (gylaz) [inline config/style_guides/thoughtbot/ruby.yml]: Same as above.

File: config/style_guides/thoughtbot/ruby.yml [hunk 19]
```diff
@@ -1,260 +1,1063 @@
...
+  StyleGuide: https://github.com/bbatsov/ruby-style-guide#utf-8
+  Enabled: false
+  EnforcedStyle: always
+  SupportedStyles:
+  - when_needed
+  - always
+Style/FileName:
+  Description: Use snake_case for source file names.
+  StyleGuide: https://github.com/bbatsov/ruby-style-guide#snake-case-files
+  Enabled: false
```
Comment (gylaz) [inline config/style_guides/thoughtbot/ruby.yml]: Enable

File: config/style_guides/thoughtbot/ruby.yml [hunk 20]
```diff
@@ -1,260 +1,1063 @@
...
+  Exclude: []
+Style/FirstParameterIndentation:
+  Description: Checks the indentation of the first parameter in a method call.
+  Enabled: false
```
Comment (gylaz) [inline config/style_guides/thoughtbot/ruby.yml]: Enable. A bit confusing, but reading into it, seems like a good practice. [Link](https://github.com/bbatsov/rubocop/blob/master/config/default.yml#L277).

**Comment by gylaz** [comment]
@jferris Where do we land on things that are a best practice in ruby, but we don't explicitly call it out in our style guide?

@salbertson There's a lot here to go through. Would it be better to generate a full config for the current thoughtbot config, and disable the rules that are not thoughtbot approved?

File: config/style_guides/thoughtbot/ruby.yml (hunk 12 above)
Comment (teoljungberg) [inline config/style_guides/thoughtbot/ruby.yml]: @gylaz our style guide says to prefer `reduce` instead of `inject`?

Comment (gylaz) [inline config/style_guides/thoughtbot/ruby.yml]: [Yep](https://github.com/thoughtbot/guides/tree/master/style/ruby#ruby)

**Comment by jferris** [comment]
> Where do we land on things that are a best practice in ruby, but we don't explicitly call it out in our style guide?

In order to be consistent, I think it would be best to leave them disabled and open a pull request to our style guides.

Thanks a lot for going through all of these. I know it's a lot of work, and not a particularly interesting challenge.

File: config/style_guides/thoughtbot/ruby.yml (hunk 1 above)
Comment (salbertson) [inline config/style_guides/thoughtbot/ruby.yml]: :+1:

File: config/style_guides/thoughtbot/ruby.yml (hunk 2 above)
Comment (salbertson) [inline config/style_guides/thoughtbot/ruby.yml]: Done

File: config/style_guides/thoughtbot/ruby.yml (hunk 3 above)
Comment (salbertson) [inline config/style_guides/thoughtbot/ruby.yml]: Done

File: config/style_guides/thoughtbot/ruby.yml (hunk 4 above)
Comment (salbertson) [inline config/style_guides/thoughtbot/ruby.yml]: Done

File: config/style_guides/thoughtbot/ruby.yml (hunk 5 above)
Comment (salbertson) [inline config/style_guides/thoughtbot/ruby.yml]: @jferris?

File: config/style_guides/thoughtbot/ruby.yml (hunk 6 above)
Comment (salbertson) [inline config/style_guides/thoughtbot/ruby.yml]: Done

File: config/style_guides/thoughtbot/ruby.yml (hunk 7 above)
Comment (salbertson) [inline config/style_guides/thoughtbot/ruby.yml]: So enable it?

File: config/style_guides/thoughtbot/ruby.yml (hunk 8 above)
Comment (salbertson) [inline config/style_guides/thoughtbot/ruby.yml]: Enable it?

File: config/style_guides/thoughtbot/ruby.yml (hunk 9 above)
Comment (salbertson) [inline config/style_guides/thoughtbot/ruby.yml]: Enable it?

File: config/style_guides/thoughtbot/ruby.yml (hunk 10 above)
Comment (salbertson) [inline config/style_guides/thoughtbot/ruby.yml]: So I'll leave it disabled.

File: config/style_guides/thoughtbot/ruby.yml (hunk 11 above)
Comment (salbertson) [inline config/style_guides/thoughtbot/ruby.yml]: Done

File: config/style_guides/thoughtbot/ruby.yml (hunk 12 above)
Comment (salbertson) [inline config/style_guides/thoughtbot/ruby.yml]: Done

File: config/style_guides/thoughtbot/ruby.yml (hunk 13 above)
Comment (salbertson) [inline config/style_guides/thoughtbot/ruby.yml]: Done

File: config/style_guides/thoughtbot/ruby.yml (hunk 14 above)
Comment (salbertson) [inline config/style_guides/thoughtbot/ruby.yml]: Done

File: config/style_guides/thoughtbot/ruby.yml (hunk 15 above)
Comment (salbertson) [inline config/style_guides/thoughtbot/ruby.yml]: Done

File: config/style_guides/thoughtbot/ruby.yml (hunk 16 above)
Comment (salbertson) [inline config/style_guides/thoughtbot/ruby.yml]: Done

File: config/style_guides/thoughtbot/ruby.yml (hunk 17 above)
Comment (salbertson) [inline config/style_guides/thoughtbot/ruby.yml]: @jferris?

File: config/style_guides/thoughtbot/ruby.yml (hunk 19 above)
Comment (salbertson) [inline config/style_guides/thoughtbot/ruby.yml]: Done

File: config/style_guides/thoughtbot/ruby.yml (hunk 20 above)
Comment (salbertson) [inline config/style_guides/thoughtbot/ruby.yml]: Done

File: config/style_guides/thoughtbot/ruby.yml [hunk 21]
```diff
@@ -1,260 +1,1063 @@
...
+    - e
+Style/SingleLineMethods:
+  Description: Avoid single-line methods.
+  StyleGuide: https://github.com/bbatsov/ruby-style-guide#no-single-line-methods
+  Enabled: false
+  AllowIfMethodIsEmpty: true
+Style/StringLiterals:
+  Description: Checks if uses of quotes match the configured preference.
+  StyleGuide: https://github.com/bbatsov/ruby-style-guide#consistent-string-literals
   Enabled: false
```
Comment (salbertson) [inline config/style_guides/thoughtbot/ruby.yml]: @gylaz, enable right?

File: config/style_guides/thoughtbot/ruby.yml (hunk 7 above)
Comment (gylaz) [inline config/style_guides/thoughtbot/ruby.yml]: I'm conflicted. I would say yes, but its not explicitly in our style guide, which @jferris said we should only enable what's in our style guide. Though if I've seen this on a PR, I would make a comment to drop the explicit braces. Should we make a PR for the tb guides?

Comment (jferris) [inline config/style_guides/thoughtbot/ruby.yml]: I think we should make pull requests to the guides for anything which we feel like should be enabled.

File: config/style_guides/thoughtbot/ruby.yml (hunk 5 above)
Comment (jferris) [inline config/style_guides/thoughtbot/ruby.yml]: Yes, this is in our guides:

> Prefer `&&` and `||` over `and` and `or`.

File: config/style_guides/thoughtbot/ruby.yml (hunk 17 above)
Comment (jferris) [inline config/style_guides/thoughtbot/ruby.yml]: We don't mention it explicitly, but our example doesn't show them: https://github.com/thoughtbot/guides/blob/master/style/ruby/sample.rb

I'd say enable and maybe submit a pull request to the guides to clarify.

File: config/style_guides/thoughtbot/ruby.yml (hunk 8 above)
Comment (gylaz) [inline config/style_guides/thoughtbot/ruby.yml]: No, it's not explicit in our guides.

File: config/style_guides/thoughtbot/ruby.yml (hunk 9 above)
Comment (gylaz) [inline config/style_guides/thoughtbot/ruby.yml]: No, it's not explicit in our guides.

File: config/style_guides/thoughtbot/ruby.yml (hunk 5 above)
Comment (salbertson) [inline config/style_guides/thoughtbot/ruby.yml]: Done.

File: config/style_guides/thoughtbot/ruby.yml (hunk 7 above)
Comment (salbertson) [inline config/style_guides/thoughtbot/ruby.yml]: Ok, I'll hold off until it's in the guide repo.

File: config/style_guides/thoughtbot/ruby.yml (hunk 17 above)
Comment (salbertson) [inline config/style_guides/thoughtbot/ruby.yml]: Done, enabled.

File: config/style_guides/thoughtbot/ruby.yml (hunk 18 above)
Comment (salbertson) [inline config/style_guides/thoughtbot/ruby.yml]: Done, enabled.

**Comment by salbertson** [comment]
I've made changes based on feedback. Here are the rules that are enabled:

```
Style/AlignHash
Style/AlignParameters
Style/AndOr
Style/BarePercentLiterals
Style/CollectionMethods
Style/DotPosition
Style/EmptyLineBetweenDefs
Style/EmptyLinesAroundBlockBody
Style/EmptyLinesAroundClassBody
Style/EmptyLinesAroundModuleBody
Style/FileName
Style/FirstParameterIndentation
Style/StringLiterals
MultilineOperationIndentation
```

File: config/style_guides/thoughtbot/ruby.yml [hunk 22]
```diff
@@ -1,264 +1,1054 @@
...
+  Max: 100
+Metrics/CyclomaticComplexity:
+  Description: A complexity metric that is strongly correlated to the number of test
+    cases needed to validate a method.
+  Enabled: false
+  Max: 6
+Metrics/LineLength:
+  Description: Limit lines to 80 characters.
+  StyleGuide: https://github.com/bbatsov/ruby-style-guide#80-character-limits
   Enabled: false
```
Comment (salbertson) [inline config/style_guides/thoughtbot/ruby.yml]: What about `LineLength`?

**Comment by salbertson** [comment]
Closing this until https://github.com/thoughtbot/guides/issues/318 is figured out.

**Pull request closed without merging**